pip/req/req_set.py

```diff
@@ -228,7 +232,11 @@ def prepare_files(self, finder):
             if not self.ignore_installed and not req_to_install.editable:
                 req_to_install.check_if_exists()
                 if req_to_install.satisfied_by:
-                    if self.upgrade:
+                    # check if we are upgrading and that we don't already have
+                    # an exact version match
+                    if self.upgrade and not (
+                            req_to_install.satisfied_by.as_requirement() ==
+                            req_to_install.req):
```

Comment:
Hmmm, I agree that it seems like that would be better, but when I tried it, it seems to break:

As-is:

```
$ pip install -U wheel==0.24.0
Requirement already satisfied (use --upgrade to upgrade): wheel==0.24.0 in /Users/marca/python/virtualenvs/pip/lib/python2.7/site-packages
```

but if I do this:

```diff
--- a/pip/req/req_set.py
+++ b/pip/req/req_set.py
@@ -230,8 +230,8 @@ class RequirementSet(object):
                 if req_to_install.satisfied_by:
                     # check if we are upgrading and that we don't already have
                     # an exact version match
-                    if self.upgrade and not (
-                            req_to_install.satisfied_by.as_requirement() ==
+                    if self.upgrade and (
+                            req_to_install.satisfied_by.as_requirement() !=
                             req_to_install.req):
                         if not self.force_reinstall and not req_to_install.url:
                             try:
```

then I get a different and incorrect behavior:

```
$ pip install -U wheel==0.24.0
processing ~/.pdbrc.py
[5] > /Users/marca/dev/git-repos/pip/pip/index.py(284)find_requirement()
-> def mkurl_pypi_url(url):
```

Note that now it's calling `PackageFinder.find_requirement`, whereas before it wasn't.

I admit to being mystified by this. Any ideas why this is happening?